### PR TITLE
Ruby Enumerable Methods

### PR DESCRIPTION
**Pull request by brenoxav**
Reimplementation of the following Ruby Enumerable methods:
- #each
- #each_with_index
- #select
- #all
- #any
- #none
- #count
- #map
- #inject

### CODE REVIEW

**Review by Georjane — CHANGES_REQUESTED** [review]
## Status: Changes Required 🛑 ❌ 
### Hi Team @brenoxav  and @cgcarlosg 🙌🙌
Nice work on the enumerable methods.👏 I see you have a basic understanding of what each of the methods is supposed to do. But now you have to implement solutions for all possible cases like when a block is given or not, and when the method is called on a range, a hash, or an array.🚫
- ### my_each method

For example:

`a = [ "a", "b", "c" ]`
`puts a.each`

returns `#<Enumerator:0x0000000002360f40>`

but your my_each method,

`puts a.my_each` 

returns the following error message


errors


- [x] Normally, Enumerables `#each` returns an Enumerator if no block is given. But your `#my_each` throws an error when no block is given. 
- [x] Enumerables` #each` when a block is given when self is a range returns the range itself after calling the given block once for each element in self but your `#my_each` throws an error when called on a range

`a = (1..5)`
`puts a.each{|x| print x, " -- " }` returns  `1 -- 2 -- 3 -- 4 -- 5 --`
but `puts a.my_each{|x| print x, " -- " }` returns an error

- [x] A similar case for Enumerables `#each` when a block is given when self is a range calls the given block once for each element in self

- [x] Normally, Enumerables` #my_each` when a block is given when self is a hash behaves like #each calls the given block once for each element in self but your method doesn't work like this.
- ### my_each_with_index method

- [x] it should return an enumerator if no block is given

- [x] when a block is given when self is a range, it should return the range itself after calling the given block once for each element in self

- [x]  when a block is given when self is a range it should behave like `#each_with_index` which calls the given block once for each element in self

- [x] when a block is given when self is a hash it should behave like `#each_with_index` which calls the given block once for each element in self

- ### my_select method

- [x] it should return an array containing all elements of enum for which the given block returns a true value::range

- ### my_all, my_any, my_none methods

- [x] it should return false if the block returns false or nil::range

- [x] when no block or argument is given it should return true when none of the collection members are false or nil

- [x] when no block or argument is given it should return false when one of the collection members is false or nil

- [x] when a class is passed as an argument it should return true if all of the collection is a member of such class::Integer, class::Numeric

- [x] when a class is passed as an argument it should return false if any of the collection is not a member of such class

- [x] when a Regex is passed as an argument it should return true if all of the collection matches the Regex

- [x] when a pattern other than Regex or a Class is given it should return false if any of the collection does not match the pattern

- ### my_count method

- [x] it should return the number of items in enum through enumeration

- [x] it should return the number of items in enum through enumeration::range

- [x] it should count the number of items in enum that are equal to the item if an argument is given

- ### my_map method
- [x] it should return a new array with the results of running a given block once for every element in enum::range

- [x] it should return an Enumerator if no block is given

- ### my_inject method
- [x]  when a block is given without an initial value it should combine all elements of enum by applying a binary operation, specified by a block::range

- [x] when a symbol is specified without an initial value it should combine each element of the collection by applying the symbol as a named method

- [x] when a symbol is specified with an initial value it should combine each element of the collection by applying the symbol as a named method::range

- ### multiply_els method

- [x] it should accept an array as an argument and multiply all the elements of the array together using #my_inject

Please go through each of the required changes carefully, fix them, and request another code review.
You can find a good resource for the enumerable methods [here](https://ruby-doc.org/core-2.6.5/Enumerable.html) 👍🏽 

Feel free to reach out to me on Slack if you have any questions, **_Witah Georjane_**🤝🤝

Happy Coding and Goodluck.👍👌💻

**Review by layzrd — CHANGES_REQUESTED** [review]
### Status: Require changes  ♲
Hi @cgcarlosg and @brenoxav, I am [Abdulaziz Redi Ali](https://azizali.ml/) your TSE in this project. 👏  ✨ I applaud you for your hard work in implementing custom enumerable methods. Most requirements have done 👌, Kindly fix the following issues.

[REQUIRED]

- [x] 1)  #my_each method should return an Enumerator if no block is given
- [x] 2) #my_each_with_index  should return an enumerator if no block is given
- [x] 3) #my_select should return an enumerator if no block is given
- [x] 4) #my_all? should return false when one of the collection members are false and when no block or argument is given 
  	E.g. `[1, false, 'hi', []].my_all?`
- [x] 5) #my_all? should return true when a class is passed as an argument and all of the collection is a member of such class     
	E.g. `[2 ,3, 4].my_all?(Numeric)`     
- [x] 6) #my_any? Should return true when no block or argument is given and at least one of the collection is not false or nil
     	E.g. `[nil, false, true, []].my_any?`
- [x] 7) #my_any? should return true when a class is passed as an argument and at least one of the collection is a member of such class
      `[2, 4, 5, 6, 7].my_any?(Numeric)`
- [x] 8) #my_none? returns false if the block ever returns true for all elements::range
     	`(10..20).none?(proc { |num| num > 19 })`
- [x] 9) #my_none? should return true none of the collection members is true
 	`[nil, false].my_none?`
- [x] 10) #my_none? should return false when a pattern other than Regex or a Class is given and only if one of the collection matches the pattern
- [x] 11) #my_count should return the number of items in enum through enumeration
       E.g. `[1, 2, 3, 4, 5].my_count`  should return 5
- [x] 12)  #my_count should return the number of items in enum through enumeration::range
      E.g `(1..5).my_count`  should return 5
- [x] 13) #my_count should count the number of items in enum that are equal to an item if an argument is given
    e.g.  `[6, 6, 2].my_count(6)`
- [ ] 14) when a symbol is specified without an initial value #my_inject should combine each element of the collection by applying the symbol as a named method 
       E.g 
 ```  
[1, 3, 6, 7].my_inject(:+)
(1..5).my_inject(:*)
 [2, 3, 6, 7].my_inject(20, :*)
 (1..5).my_inject(2, :*)
```
- [x] 15) Put the #multiply_els method outside of the Enumerable module
- [x] 16) Add descriptive content and setup instruction in the readme file

[OPTIONAL]

 - [ ] Enable all rubocop rules

Once you fix the issues, kindly submit another code review. ♲

Happy coding 👨‍💻
[Abdulaziz Redi Ali](https://azizali.ml/) | [Contact me on Slack](https://microverse-students.slack.com/team/U011SD6JUH2)

File: enumerables.rb [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    block_given? ? my_each { |e| arr.push(e) if yield(e) } : (return to_enum)
+    arr
+  end
+
+  def my_all?(arg = nil)
+    my_each do |e|
+      if block_given?
+        return false unless yield(e)
+      elsif !block_given? && !arg
+        return false
```
Comment (ebeagusamuel) [inline enumerables.rb]: - [x]  when no block or argument is given, your my_all? should return true when none of the collection members are false or nil. Your code is currently returning false if no block or argument is given.

File: enumerables.rb [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+      elsif arg.instance_of?(Class)
+        return true unless e.is_a? arg
+      elsif arg
+        return true unless e == arg
+      else
+        return false
+      end
+    end
+    false
+  end
```
Comment (ebeagusamuel) [inline enumerables.rb]: - [x] your my_none? should return true if the block never returns true for all elements.
- [x] when no block or argument is given, your my_none? should return false only if one of the collection members is true.
- [x] when a pattern other than Regex or a Class is given, your my_none? should return false only if one of the collection matches the pattern.

File: enumerables.rb [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        acc = e
+      elsif acc.instance_of?(Range)
+        my_each { |i| acc += i }
+      else
+        acc = yield(acc, e)
+      end
+    end
+    acc
+  end
+end
```
Comment (ebeagusamuel) [inline enumerables.rb]: - [x] when a symbol is specified without an initial value, your my_inject should combine each element of the collection by applying the symbol as a named method. i.e `[1, 2, 3, 4].my_inject(:+)` should return 10.
- [x]  when a symbol is specified without an initial value, your my_inject should combine each element of the collection by applying the symbol as a named method: range. i.e it should be able to work also on a range(`(1..4).my_inject(:+)` should return 10)
- [x] when a symbol is specified with an initial value, my_inject combines each element of the collection by applying the symbol as a named method. i.e `[1, 2, 3, 4].my_inject(20, :*)` should return 480 and should also be able to work on a range too.

**Review by ebeagusamuel — CHANGES_REQUESTED** [review]
# Changes Required :recycle: 
Hello team :wave: 
Nice work you've done so far. Just a few changes to be made to get the project approved.
Please check the comment below and also update your PR message with more information.
Happy coding :+1: 
slack(@ebeagusamuel)

**Review by layzrd — CHANGES_REQUESTED** [review]
### Status: Require changes  ♲
Hi @cgcarlosg and @brenoxav, I am [Abdulaziz Redi Ali](https://azizali.ml/), your TSE in this PR. 👏 👏 You made good progress on this project. ✨ I applaud you for your hard work in implementing custom enumerable methods. Most requirements have done 👌, Kindly fix the following issues.

[REQUIRED]

- [ ] If no block or argument is given `#my_all?` should return false when one of the collection members is false or nil
_Example:_
`[nil, true, 99].all?  #=> false` so, `#my_all?` should return the same result as the original method
`[nil, true, 99].my_all?   #=> false`

- [ ] If a block is given with an initial value `#my_inject` should combine all elements of enum by applying a binary operation, specified by a block

   Example:
   `(2..4).inject(2) { |product, n| product * n } #=> 48`

- [ ] If a symbol is specified with an initial value #my_inject should combine each element of the collection by applying the symbol as a named method.

Example using array:
`[2, 3, 4].inject(2, :*)  #=> 48 so`, `#my_inject` should return the same result as the original method
`[2, 3, 4].inject(2, :*)   #=> 48`

Example using range:
`(2..4).inject(2, :*)  #=> 48` so, `#my_inject` should return the same result as the original method
`(2..4).inject(2, :*)   #=> 48`

[OPTIONAL]

- [ ] Enable all rubocop rules
- [ ] Refactor your complex methods(my_all?, my_none?, my_all?) by extracting a code fragment into a new method

Once you fix the issues, kindly submit another code review. ♲

Happy coding 👨‍💻
[Abdulaziz Redi Ali](https://azizali.ml/) | [Connect me on Slack](https://microverse-students.slack.com/team/U011SD6JUH2)

**Review by ElijahTobs — APPROVED** [review]
# Status: Approved! :tada: 

Congratulations! @brenoxav and @cgcarlosg 

 Good job implementing the change as requested by the previous TSE.

## [NICE TO HAVE]
- [ ] Add a live demo link of your methods to the README file. You can use [repl](https://repl.it/)
- [x] Fix the issues link in your README file.
- [x] List the enumerables methods in your PR message and README

## You can now merge and submit.

## Happy Coding :computer: 

_Feel free to reach out to me on slack_ **`@Elijah Ayandokun`**
  **Elijah O. Ayandokun**
[Github](https://github.com/ElijahTobs) | [LinkedIn](https://www.linkedin.com/in/ayandokunelijah/) | [Twitter](https://twitter.com/ElijahTobs)

**Pull request merged**